File: examples/Csharp.ExampleApplication/Hooks/PostPublish/BulkLoggingHook.cs
using System.Threading;
using System.Threading.Tasks;
using Microsoft.Extensions.Logging;
using Tableau.Migration.Engine.Hooks.PostPublish;

namespace Csharp.ExampleApplication.Hooks.PostPublish
{
    #region class
    public class BulkLoggingHook<T> : BulkPostPublishHookBase<T>
    {
        private readonly ILogger<BulkLoggingHook<T>> _logger;

        public BulkLoggingHook(ILogger<BulkLoggingHook<T>> logger)
        {
            _logger = logger;
        }

        public override Task<BulkPostPublishContext<T>?> ExecuteAsync(BulkPostPublishContext<T> ctx, CancellationToken cancel)
        {
            // Log the number of items published in the batch.
            _logger.LogInformation(
                "Published {Count} {ContentType} item(s).",
                ctx.PublishedItems.Count,
                typeof(T).Name);

            return Task.FromResult<BulkPostPublishContext<T>?>(ctx);
        }
    }
    #endregion
}

File: examples/Python.ExampleApplication/Hooks/post_publish/bulk_logging_hook.py
import logging
from tableau_migration import (
    BulkPostPublishHookBase,
    BulkPostPublishContext,
    IDataSource)


class BulkLoggingHookForDataSources(BulkPostPublishHookBase[IDataSource]):
    def __init__(self) -> None:
        super().__init__()
        
        # Create a logger for this class
        self._logger = logging.getLogger(__name__)
        
    def execute(self, ctx: BulkPostPublishContext[IDataSource]) -> BulkPostPublishContext[IDataSource]:
        # Log the number of items published in the batch.
        self._logger.info("Published %d IDataSource item(s).", ctx.published_items.count)
        return None
    